File: ios/Tests/SentinelSDKTests/SentinelSDKTests.swift
import XCTest
@testable import SentinelSDK

final class SentinelSDKTests: XCTestCase {

    // MARK: - Personal Data Validation Tests

    func testPersonalDataValidationSuccess() {
        let validator = PersonalDataValidator()
        let input = PersonalDataValidator.Input(
            fullName: "Carlos Eduardo da Silva",
            documentId: "123.456.789-09", // Valid CPF check digits
            email: "carlos@example.com",
            phoneNumber: "+5511999998888"
        )
        let result = validator.validate(input: input)

        XCTAssertEqual(result.fullName, "Carlos Eduardo da Silva")
        XCTAssertTrue(result.validation.isValid)
        XCTAssertTrue(result.validation.nameFormatValid)
        XCTAssertTrue(result.validation.documentValid)
        XCTAssertTrue(result.validation.phoneFormatValid)
        XCTAssertTrue(result.validation.validationErrors.isEmpty)
    }

    func testDemoAppPresetsValidation() {
        let validator = PersonalDataValidator()

        // 1. Valid Preset used in Demo App
        let validPreset = PersonalDataValidator.Input(
            fullName: "Carlos Eduardo da Silva",
            documentId: "123.456.789-09",
            email: "carlos.silva@example.com",
            phoneNumber: "+5511999998888"
        )
        let validResult = validator.validate(input: validPreset)
        XCTAssertTrue(validResult.validation.isValid)
        XCTAssertTrue(validResult.validation.validationErrors.isEmpty)

        // 2. Invalid Preset used in Demo App
        let invalidPreset = PersonalDataValidator.Input(
            fullName: "Carlos",
            documentId: "111.111.111-11",
            email: "invalid-email-address",
            phoneNumber: "1234"
        )
        let invalidResult = validator.validate(input: invalidPreset)
        XCTAssertFalse(invalidResult.validation.isValid)
        XCTAssertFalse(invalidResult.validation.nameFormatValid)
        XCTAssertFalse(invalidResult.validation.documentValid)
        XCTAssertFalse(invalidResult.validation.phoneFormatValid)
        XCTAssertEqual(invalidResult.validation.validationErrors.count, 4)
    }

    func testPersonalDataValidationCPFAlgorithm() {
        let validator = PersonalDataValidator()

        // Valid CPFs
        XCTAssertTrue(validator.validateCPF("12345678909"))
        XCTAssertTrue(validator.validateCPF("52998224725"))

        // Invalid CPFs
        XCTAssertFalse(validator.validateCPF("12345678900"))
        XCTAssertFalse(validator.validateCPF("11111111111"))
        XCTAssertFalse(validator.validateCPF("00000000000"))
        XCTAssertFalse(validator.validateCPF("12345"))
    }

    func testPersonalDataValidationCNPJAlgorithm() {
        let validator = PersonalDataValidator()

        // Valid CNPJs
        XCTAssertTrue(validator.validateCNPJ("00000000000191"))
        XCTAssertTrue(validator.validateCNPJ("11222333000181"))

        // Invalid CNPJs
        XCTAssertFalse(validator.validateCNPJ("00000000000100"))
        XCTAssertFalse(validator.validateCNPJ("11111111111111"))
        XCTAssertFalse(validator.validateCNPJ("1234"))
    }

    func testPersonalDataValidationFailures() {
        let validator = PersonalDataValidator()
        let input = PersonalDataValidator.Input(
            fullName: "Carlos", // Single token missing last name
            documentId: "12345678900", // Invalid check digits
            email: "invalid-email-address",
            phoneNumber: "123" // Too short
        )
        let result = validator.validate(input: input)

        XCTAssertFalse(result.validation.isValid)
        XCTAssertFalse(result.validation.nameFormatValid)
        XCTAssertFalse(result.validation.documentValid)
        XCTAssertFalse(result.validation.phoneFormatValid)
        XCTAssertEqual(result.validation.validationErrors.count, 4)
    }

    // MARK: - Device Telemetry Tests

    func testDeviceModelMapping() {
        let collector = DeviceDataCollector()
        XCTAssertEqual(collector.mapModelIdentifierToMarketingName("iPhone16,1"), "iPhone 15 Pro")
        XCTAssertEqual(collector.mapModelIdentifierToMarketingName("iPhone16,2"), "iPhone 15 Pro Max")
        XCTAssertEqual(collector.mapModelIdentifierToMarketingName("iPhone15,2"), "iPhone 14 Pro")
        XCTAssertEqual(collector.mapModelIdentifierToMarketingName("iPhone14,2"), "iPhone 13 Pro")
        XCTAssertEqual(collector.mapModelIdentifierToMarketingName("CustomDevice"), "CustomDevice")
    }

    // MARK: - Cryptographic Signing Tests

    func testPayloadSignerOutputAndVerification() {
        let secretKey = "enterprise_secret_key_12345"
        let signer = PayloadSigner(secretKey: secretKey)
        let testData = "test_payload_sample_content".data(using: .utf8)!
        let nonce = "unique_session_nonce_abc"

        let securityData = signer.sign(data: testData, nonce: nonce)

        XCTAssertFalse(securityData.payloadHash.isEmpty)
        XCTAssertFalse(securityData.signature.isEmpty)
        XCTAssertEqual(securityData.nonce, nonce)
        XCTAssertFalse(securityData.tamperDetected)

        // Verify signature matches
        XCTAssertTrue(signer.verify(payloadHash: securityData.payloadHash, nonce: nonce, signature: securityData.signature))

        // Tampered hash must fail verification
        XCTAssertFalse(signer.verify(payloadHash: "tampered_hash", nonce: nonce, signature: securityData.signature))
    }

    // MARK: - App Detection Collector Tests

    func testAppDetectionCollectorDefaults() async {
        let collector = AppDetectionCollector()
        let result = await MainActor.run {
            collector.scan()
        }

        XCTAssertEqual(result.scanStrategy, "targeted_scheme_queries")
        XCTAssertGreaterThanOrEqual(result.totalTargetsScanned, 20)
        XCTAssertGreaterThan(result.unresolvedSchemesCount, 0)
    }

    func testAppDetectionCollectorCustomTargets() async {
        let collector = AppDetectionCollector()
        let customTargets = [
            AppDetectionCollector.TargetApp(id: "custom_bet", name: "Custom Bet", scheme: "custombet"),
            AppDetectionCollector.TargetApp(id: "no_scheme_app", name: "No Scheme", scheme: nil)
        ]

        let result = await MainActor.run {
            collector.scan(targets: customTargets)
        }

        XCTAssertEqual(result.totalTargetsScanned, 2)
        XCTAssertEqual(result.unresolvedSchemesCount, 1)
    }

    // MARK: - End-to-End SDK Facade Tests

    func testSDKCaptureJSONOutput() async throws {
        let sdk = SentinelSDK.shared
        sdk.initialize(apiKey: "test_enterprise_api_key_12345", environment: "sandbox")

        let config = SentinelSDK.CaptureConfiguration(
            userData: PersonalDataValidator.Input(
                fullName: "Carlos Eduardo da Silva",
                documentId: "123.456.789-09",
                email: "carlos.silva@example.com",
                phoneNumber: "+5511999998888"
            ),
            locationTimeoutSeconds: 1.0,
            wrapper: "native_ios_demo"
        )

        let jsonString = try await sdk.captureJSON(configuration: config)

        XCTAssertFalse(jsonString.isEmpty)
        XCTAssertTrue(jsonString.contains("\"sdk_version\" : \"1.0.0\""))
        XCTAssertTrue(jsonString.contains("\"full_name\" : \"Carlos Eduardo da Silva\""))
        XCTAssertTrue(jsonString.contains("\"wrapper\" : \"native_ios_demo\""))
        XCTAssertTrue(jsonString.contains("\"payload_hash\""))
        XCTAssertTrue(jsonString.contains("\"signature\""))
    }
}

File: android/sentinel-core/src/test/kotlin/com/sentinel/sdk/SentinelSDKTests.kt
package com.sentinel.sdk

import com.sentinel.sdk.collectors.AppDetectionCollector
import com.sentinel.sdk.collectors.PersonalDataValidator
import com.sentinel.sdk.models.SentinelCoordinates
import com.sentinel.sdk.models.SentinelDeviceData
import com.sentinel.sdk.models.SentinelInstalledApps
import com.sentinel.sdk.models.SentinelLocationData
import com.sentinel.sdk.models.SentinelMetadata
import com.sentinel.sdk.models.SentinelPayload
import com.sentinel.sdk.models.SentinelSecurityData
import com.sentinel.sdk.models.SentinelTelephonyData
import com.sentinel.sdk.models.SentinelUserData
import com.sentinel.sdk.models.SentinelUserValidation
import com.sentinel.sdk.security.PayloadSigner
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

class SentinelSDKTests {

    // MARK: - Personal Data Validation Tests

    @Test
    fun testPersonalDataValidationSuccess() {
        val validator = PersonalDataValidator()
        val input = PersonalDataValidator.Input(
            fullName = "Carlos Eduardo da Silva",
            documentId = "123.456.789-09", // Valid CPF check digits
            email = "carlos@example.com",
            phoneNumber = "+5511999998888"
        )
        val result = validator.validate(input)

        assertEquals("Carlos Eduardo da Silva", result.fullName)
        assertTrue(result.validation.isValid)
        assertTrue(result.validation.nameFormatValid)
        assertTrue(result.validation.documentValid)
        assertTrue(result.validation.phoneFormatValid)
        assertTrue(result.validation.validationErrors.isEmpty())
    }

    @Test
    fun testPersonalDataValidationCPFAlgorithm() {
        val validator = PersonalDataValidator()

        // Valid CPFs
        assertTrue(validator.validateCPF("12345678909"))
        assertTrue(validator.validateCPF("52998224725"))

        // Invalid CPFs
        assertFalse(validator.validateCPF("12345678900"))
        assertFalse(validator.validateCPF("11111111111"))
        assertFalse(validator.validateCPF("00000000000"))
        assertFalse(validator.validateCPF("12345"))
    }

    @Test
    fun testPersonalDataValidationCNPJAlgorithm() {
        val validator = PersonalDataValidator()

        // Valid CNPJs
        assertTrue(validator.validateCNPJ("00000000000191"))
        assertTrue(validator.validateCNPJ("11222333000181"))

        // Invalid CNPJs
        assertFalse(validator.validateCNPJ("00000000000100"))
        assertFalse(validator.validateCNPJ("11111111111111"))
        assertFalse(validator.validateCNPJ("1234"))
    }

    @Test
    fun testPersonalDataValidationFailures() {
        val validator = PersonalDataValidator()
        val input = PersonalDataValidator.Input(
            fullName = "Carlos", // Missing last name
            documentId = "12345678900", // Invalid check digits
            email = "invalid-email-address",
            phoneNumber = "123" // Too short
        )
        val result = validator.validate(input)

        assertFalse(result.validation.isValid)
        assertFalse(result.validation.nameFormatValid)
        assertFalse(result.validation.documentValid)
        assertFalse(result.validation.phoneFormatValid)
        assertEquals(4, result.validation.validationErrors.size)
    }

    // MARK: - Cryptographic Signing Tests

    @Test
    fun testPayloadSignerOutputAndVerification() {
        val secretKey = "enterprise_secret_key_12345"
        val signer = PayloadSigner(secretKey)
        val testPayload = "{\"test\": \"payload_content\"}"
        val nonce = "session_nonce_xyz"

        val securityData = signer.sign(testPayload, nonce)

        assertTrue(securityData.payloadHash.isNotEmpty())
        assertTrue(securityData.signature.isNotEmpty())
        assertEquals(nonce, securityData.nonce)
        assertFalse(securityData.tamperDetected)

        // Verify signature matches
        assertTrue(signer.verify(securityData.payloadHash, nonce, securityData.signature))

        // Tampered hash must fail verification
        assertFalse(signer.verify("tampered_payload_hash", nonce, securityData.signature))
    }

    // MARK: - App Detection Targets Tests

    @Test
    fun testAppDetectionDefaultTargets() {
        val targets = AppDetectionCollector.DEFAULT_TARGETS
        assertTrue(targets.size >= 20)
        assertTrue(targets.any { it.packageName == "com.bet365.app" })
        assertTrue(targets.any { it.packageName == "com.superbet.app" })
        assertTrue(targets.any { it.packageName == "com.sportybet.android" })
        assertTrue(targets.any { it.packageName == "com.betano.app" })
        assertTrue(targets.any { it.packageName == "gr.novibet" })
        assertTrue(targets.any { it.packageName == "com.sportingbet.sportsbook" })
    }

    @Test
    fun testDynamicAppTargetParsingUserRequestedApps() {
        val input = "bet365, Superbet, SportyBet, Betano"
        val targets = AppDetectionCollector.parseTargetsFromInput(input)

        assertTrue(targets.size >= 4)
        assertTrue(targets.any { it.packageName == "com.bet365.app" })
        assertTrue(targets.any { it.packageName == "com.superbet.app" })
        assertTrue(targets.any { it.packageName == "com.sportybet.android" })
        assertTrue(targets.any { it.packageName == "com.betano.app" })
    }

    @Test
    fun testDynamicAppTargetParsingCustomFormats() {
        val input = "com.custom.app, MyBettingApp:com.mybet.pkg, CustomSocial (com.social.app)"
        val targets = AppDetectionCollector.parseTargetsFromInput(input)

        assertEquals(3, targets.size)
        assertTrue(targets.any { it.packageName == "com.custom.app" })
        assertTrue(targets.any { it.packageName == "com.mybet.pkg" && it.name == "MyBettingApp" })
        assertTrue(targets.any { it.packageName == "com.social.app" && it.name == "CustomSocial" })
    }

    // MARK: - JSON Serialization Tests

    @Test
    fun testPayloadJsonSerialization() {
        val payload = SentinelPayload(
            metadata = SentinelMetadata(
                osVersion = "14.0",
                sessionId = "test-session-123",
                requestId = "req-123",
                timestamp = "2026-09-21T23:00:00.000Z",
                captureDurationMs = 150
            ),
            userData = SentinelUserData(
                fullName = "Maria Oliveira",
                documentId = "52998224725",
                email = "maria@example.com",
                phoneNumber = "+5511988887777",
                validation = SentinelUserValidation(
                    isValid = true,
                    nameFormatValid = true,
                    documentValid = true,
                    phoneFormatValid = true,
                    validationErrors = emptyList()
                )
            ),
            deviceData = SentinelDeviceData(
                deviceIdentifier = "test-device-uuid",
                model = "Pixel 8 Pro",
                manufacturer = "Google",
                brand = "google",
                locale = "pt_BR",
                timezone = "America/Sao_Paulo",
                screenResolution = "1344x2992",
                batteryLevel = 0.95f,
                batteryState = "charging",
                isJailbrokenOrRooted = false,
                isEmulator = false,
                telephony = SentinelTelephonyData(
                    carrierName = "Claro BR",
                    mobileCountryCode = "724",
                    mobileNetworkCode = "05",
                    isoCountryCode = "br",
                    networkType = "CELLULAR_5G",
                    isSimReady = true,
                    simOperatorName = "Claro"
                )
            ),
            locationData = SentinelLocationData(
                status = "SUCCESS",
                permissionStatus = "granted_fine",
                coordinates = SentinelCoordinates(
                    latitude = -23.55052,
                    longitude = -46.633308,
                    accuracyMeters = 3.5,
                    altitudeMeters = 760.0
                ),
                timestamp = "2026-09-21T23:00:00.000Z",
                isMockLocation = false
            ),
            installedApps = SentinelInstalledApps(
                scanStrategy = "targeted_manifest_queries",
                totalTargetsScanned = 25,
                totalDetected = 0,
                riskLevel = "CLEAN",
                hasBettingApps = false,
                detectedApps = emptyList()
            ),
            security = SentinelSecurityData(
                payloadHash = "mock_hash",
                signature = "mock_sig",
                nonce = "mock_nonce",
                tamperDetected = false
            )
        )

        val serialized = SentinelSDK.prettyJson.encodeToString(payload)

        assertTrue(serialized.contains("\"sdk_version\": \"1.0.0\""))
        assertTrue(serialized.contains("\"platform\": \"android\""))
        assertTrue(serialized.contains("\"full_name\": \"Maria Oliveira\""))
        assertTrue(serialized.contains("\"carrier_name\": \"Claro BR\""))
    }

    @Test
    fun testPayloadSignerTamperFlag() {
        val signer = PayloadSigner("secret_key_123")
        val securityData = signer.sign("{\"data\":\"val\"}", tamperDetected = true)
        assertTrue(securityData.tamperDetected)
        assertTrue(signer.verify(securityData.payloadHash, securityData.nonce, securityData.signature))
    }
}

